server/src/main/java/org/opensearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -465,17 +466,42 @@ public Builder initializeAsRemoteStoreRestore(
             }
             for (int shardNumber = 0; shardNumber < indexMetadata.getNumberOfShards(); shardNumber++) {
                 ShardId shardId = new ShardId(index, shardNumber);
+                if (indexShardRoutingTableMap.containsKey(shardId) == false) {
```

Comment:
Agree, it should always present. This check is a bit defensive as `indexShardRoutingTableMap` is a map and **IF** the key is missing, it will throw NPE.